evm/src/poseidon/poseidon_stark.rs

```diff
@@ -0,0 +1,960 @@
+
+    res.push(Column::linear_combination_with_constant(
+        [
+            (cols.virt, F::ONE),
+            (cols.already_absorbed_elements, F::ONE),
+        ],
+        F::from_canonical_usize(i),
+    ));
+
+    res.push(Column::single(cols.input[i]));
+    res.extend((1..8).map(|_| Column::zero()));
+
+    res.push(Column::single(cols.timestamp));
+
+    assert_eq!(
+        res.len(),
+        crate::memory::memory_stark::ctl_data::<F>().len()
+    );
+
+    res
+}
+
+pub fn ctl_looking_memory_filter<F: Field>(i: usize) -> Column<F> {
+    let cols = POSEIDON_COL_MAP;
+    if i == POSEIDON_SPONGE_RATE - 1 {
+        Column::single(cols.is_full_input_block)
+    } else {
+        Column::sum(once(&cols.is_full_input_block).chain(&cols.is_final_input_len[i + 1..]))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct PoseidonOp {
+    /// The base address at which inputs are read.
+    pub(crate) base_address: MemoryAddress,
+
+    /// The timestamp at which inputs are read.
+    pub(crate) timestamp: usize,
+
+    /// The input that was read. We assume that it was
+    /// previously padded.
+    pub(crate) input: Vec<u32>,
+
+    /// Length of the input before paddding.
+    pub(crate) len: usize,
+}
+
+#[derive(Copy, Clone, Default)]
+pub struct PoseidonStark<F, const D: usize> {
+    pub(crate) f: PhantomData<F>,
+}
+
+/// Information about a Poseidon operation needed for witness generation.
+impl<F: RichField + Extendable<D>, const D: usize> PoseidonStark<F, D> {
+    /// Generate the rows of the trace. Note that this does not generate the permuted columns used
+    /// in our lookup arguments, as those are computed after transposing to column-wise form.
+    fn generate_trace_rows(
+        &self,
+        operations: Vec<PoseidonOp>,
+        min_rows: usize,
+    ) -> Vec<[F; NUM_COLUMNS]> {
+        let base_len: usize = operations
+            .iter()
+            .map(|op| {
+                debug_assert!(op.input.len() % POSEIDON_SPONGE_RATE == 0);
+                op.input.len() / POSEIDON_SPONGE_RATE
+            })
+            .sum();
+
+        let num_rows = base_len.max(min_rows).next_power_of_two();
+        let mut rows = Vec::with_capacity(base_len.max(min_rows));
+
+        for op in operations {
+            rows.extend(self.generate_rows_for_op(op));
+        }
+
+        // We generate "actual" rows for padding to avoid having to store
+        // another power of x, on top of x^3 and x^6.
+        let padding_row: [F; NUM_COLUMNS] = {
+            let mut tmp_row = PoseidonColumnsView::default();
+            let padding_inp = [F::ZERO; POSEIDON_SPONGE_WIDTH];
+            Self::generate_perm(&mut tmp_row, padding_inp);
+            tmp_row
+        }
+        .into();
+        while rows.len() < num_rows {
+            rows.push(padding_row);
+        }
+        rows
+    }
+
+    fn generate_rows_for_op(&self, op: PoseidonOp) -> Vec<[F; NUM_COLUMNS]> {
```

Comment:
nit: Could just return an iterator of `[F; NUM_COLUMNS]` instead of allocating a vector.

Reply:
I tried changing this to return an iterator, but I got an error because we are taking the inputs by reference (and therefore the compiler wants me to collect the `map` into a vector anyway)... I might be doing it wrong so please let me know if you had way of doing this in mind. I've left this as is for now since it's done the same way in `KeccakSpongeStark`.

Reply:
Ah ok, no problem then! Shouldn't be an issue.